### PR TITLE
Release notes draft v1.20.0 rc.0

### PR DESCRIPTION
**Pull request by wilsonehusin**
#### What type of PR is this:

/kind documentation

#### What this PR does / why we need it:

Release notes for Kubernetes 1.20, based on `v1.20.0-rc.0` tag

#### Which issue(s) this PR fixes:

None

#### Special notes for your reviewer:

I will go through this again tomorrow with the community documentation guide in hand, too tired now and just want to get this out first :smile:

/cc @celestehorgan @tanjacky @soniasingla @JamesLaverack @puerco

### CODE REVIEW

**Review by soniasingla — APPROVED** [review]
Awesome, great work :)

Thanks @wilsonehusin 🎉

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Comment by wilsonehusin** [comment]
err I'm gonna assume that myself haven't approved this and let morning-me do the approval

/hold

**Comment by puerco** [comment]
Great job @wilsonehusin  It also looks good to me :+1: 
/lgtm

File: releases/release-1.20/release-notes-draft.md [hunk 1]
```diff
@@ -1,178 +1,256 @@
...
+
+### PID Limits graduates to General Availability
+After being enabled-by-default for a year, SIG Node graduates PID Limits to GA on both `SupportNodePidsLimit` (node-to-pod PID isolation) and `SupportPodPidsLimit` (ability to limit PIDs per pod).
+
+### API Priority and Fairness graduates to Beta
+Initially introduced in 1.18, Kubernetes 1.20 now enables API Priority and Fairness (APF) by default. This allows `kube-apiserver` to [categorize incoming requests by priority levels](https://docs.k8s.io/concepts/cluster-administration/flow-control/).
+
+### IPv4/IPv6 dual-stack
+IPv4/IPv6 dual-stack has been reimplemented to support dual-stack Services based on user and community feedback. This allows both IPv4 and IPv6 Service cluster-IP addresses to be assigned to a single service and also enables Service to be upgraded/downgraded from single stack to dual-stack and vice versa.
```
Comment (bridgetkromhout) [inline releases/release-1.20/release-notes-draft.md:25]: Hi, apologies for the lateness - I know @lachie83 provided this existing blurb, but he approves this more detailed version that @jeremyrickard asked me for whilst I was out of office. I know this is beyond last-minute but hopefully can still go in for the release notes (and please let me know if this should go somewhere else):

For reading:

### IPv4/IPv6 dual-stack
IPv4/IPv6 dual-stack has been reimplemented for 1.20 to support dual-stack Services, based on user and community feedback. If your cluster has dual-stack enabled, you can create Services which can use IPv4, IPv6, or both, and you can change this setting for existing Services. Details are available in updated [IPv4/IPv6 dual-stack docs](https://kubernetes.io/docs/concepts/services-networking/dual-stack/), which cover the nuanced array of options.

We expect this implementation to progress to beta and GA in coming releases, so we’re eager to have you comment about your dual-stack experiences in [#k8s-dual-stack](https://kubernetes.slack.com/messages/k8s-dual-stack) or in [enhancements #563](https://features.k8s.io/563). 

For copying:

```
### IPv4/IPv6 dual-stack
IPv4/IPv6 dual-stack has been reimplemented for 1.20 to support dual-stack Services, based on user and community feedback. If your cluster has dual-stack enabled, you can create Services which can use IPv4, IPv6, or both, and you can change this setting for existing Services. Details are available in updated [IPv4/IPv6 dual-stack docs](https://kubernetes.io/docs/concepts/services-networking/dual-stack/), which cover the nuanced array of options.

We expect this implementation to progress to beta and GA in coming releases, so we’re eager to have you comment about your dual-stack experiences in [#k8s-dual-stack](https://kubernetes.slack.com/messages/k8s-dual-stack) or in [enhancements #563](https://features.k8s.io/563). 
```

File: releases/release-1.20/release-notes-draft.md [hunk 2]
```diff
@@ -1,178 +1,258 @@
...
+
+### API Priority and Fairness graduates to Beta
+Initially introduced in 1.18, Kubernetes 1.20 now enables API Priority and Fairness (APF) by default. This allows `kube-apiserver` to [categorize incoming requests by priority levels](https://docs.k8s.io/concepts/cluster-administration/flow-control/).
+
+### IPv4/IPv6 dual-stack
+IPv4/IPv6 dual-stack has been reimplemented for 1.20 to support dual-stack Services, based on user and community feedback. If your cluster has dual-stack enabled, you can create Services which can use IPv4, IPv6, or both, and you can change this setting for existing Services. Details are available in updated [IPv4/IPv6 dual-stack docs](https://docs.k8s.io/concepts/services-networking/dual-stack/), which cover the nuanced array of options.
+
+We expect this implementation to progress to beta and GA in coming releases, so we’re eager to have you comment about your dual-stack experiences in [#k8s-dual-stack](https://kubernetes.slack.com/messages/k8s-dual-stack) or in [enhancements #563](https://features.k8s.io/563).
```
Comment (wilsonehusin) [inline releases/release-1.20/release-notes-draft.md:27]: @bridgetkromhout 👍🏼 👎🏼 ?

File: releases/release-1.20/release-notes-draft.md (hunk 1 above)
Comment (wilsonehusin) [reply releases/release-1.20/release-notes-draft.md]: https://github.com/kubernetes/sig-release/pull/1371#discussion_r535536114

File: releases/release-1.20/release-notes-draft.md (hunk 2 above)
Comment (bridgetkromhout) [reply releases/release-1.20/release-notes-draft.md:27]: On further review, I'd change "We expect this implementation" to "We expect this alpha implementation" just to be extra-clear?

Comment (wilsonehusin) [reply releases/release-1.20/release-notes-draft.md:27]: hmm I'm not a trained technical writer, but to me "this alpha implementation" seem to imply that there won't be any changes until they graduate -- is that accurate?

alternatively, I was thinking more along the lines of "We expect his implementation approach to progress from alpha to beta and GA"

thoughts?

Comment (bridgetkromhout) [reply releases/release-1.20/release-notes-draft.md:27]: Ah, sure! Okay, what if we changed "We expect this implementation to progress to beta and GA in coming releases" to:
```
We expect this implementation to progress from alpha to beta and GA in coming releases
```

basically I want the term alpha in there somewhere, for people who are looking for it.

Comment (bridgetkromhout) [reply releases/release-1.20/release-notes-draft.md:27]: If you don't think it works there, then a possible other place it could go is:
"IPv4/IPv6 dual-stack has been reimplemented for"

changing to "IPv4/IPv6 dual-stack has been reimplemented as alpha for"

Comment (wilsonehusin) [reply releases/release-1.20/release-notes-draft.md:27]: I like the first one! WIP...

File: releases/release-1.20/release-notes-draft.md [hunk 3]
```diff
@@ -24,7 +24,7 @@ Initially introduced in 1.18, Kubernetes 1.20 now enables API Priority and Fairn
 ### IPv4/IPv6 dual-stack
 IPv4/IPv6 dual-stack has been reimplemented for 1.20 to support dual-stack Services, based on user and community feedback. If your cluster has dual-stack enabled, you can create Services which can use IPv4, IPv6, or both, and you can change this setting for existing Services. Details are available in updated [IPv4/IPv6 dual-stack docs](https://docs.k8s.io/concepts/services-networking/dual-stack/), which cover the nuanced array of options.
 
-We expect this implementation to progress to beta and GA in coming releases, so we’re eager to have you comment about your dual-stack experiences in [#k8s-dual-stack](https://kubernetes.slack.com/messages/k8s-dual-stack) or in [enhancements #563](https://features.k8s.io/563). 
+We expect this implementation to progress from alpha to beta and GA in coming releases, so we’re eager to have you comment about your dual-stack experiences in [#k8s-dual-stack](https://kubernetes.slack.com/messages/k8s-dual-stack) or in [enhancements #563](https://features.k8s.io/563).
```
Comment (wilsonehusin) [inline releases/release-1.20/release-notes-draft.md:27]: @bridgetkromhout sounds good?

Comment (bridgetkromhout) [reply releases/release-1.20/release-notes-draft.md:27]: Great! Thanks.

File: releases/release-1.20/release-notes-draft.md [hunk 4]
```diff
@@ -1,178 +1,258 @@
...
 - Move configurable fsgroup change policy for pods to beta ([#96376](https://github.com/kubernetes/kubernetes/pull/96376), [@gnufied](https://github.com/gnufied)) [SIG Apps and Storage]
-- New flag is introduced to kubelet, i.e. `--topology-manager-scope=container|pod`.
+- New flag is introduced, i.e. --topology-manager-scope=container|pod. 
   The default value is the "container" scope. ([#92967](https://github.com/kubernetes/kubernetes/pull/92967), [@cezaryzukowski](https://github.com/cezaryzukowski)) [SIG Instrumentation, Node and Testing]
 - New parameter `defaultingType` for `PodTopologySpread` plugin allows to use k8s defined or user provided default constraints ([#95048](https://github.com/kubernetes/kubernetes/pull/95048), [@alculquicondor](https://github.com/alculquicondor)) [SIG Scheduling]
 - NodeAffinity plugin can be configured with AddedAffinity. ([#96202](https://github.com/kubernetes/kubernetes/pull/96202), [@alculquicondor](https://github.com/alculquicondor)) [SIG Node, Scheduling and Testing]
 - Promote RuntimeClass feature to GA.
```
Comment (SergeyKanzhelev) [inline releases/release-1.20/release-notes-draft.md:182]: For this one, do we need to add some action items for users to start using the v1 API?

Comment (SergeyKanzhelev) [reply releases/release-1.20/release-notes-draft.md:182]: or the fact it is in the API change section is sufficient?

Comment (wilsonehusin) [reply releases/release-1.20/release-notes-draft.md:182]: huh, I think we might need to -- what happens if users are still using pre-v1 API? is it going to break or just deprecation warning?

File: releases/release-1.20/release-notes-draft.md [hunk 5]
```diff
@@ -1,178 +1,258 @@
...
+### `TokenRequest` / `TokenRequestProjection` graduates to General Availability
+Service account tokens bound to pod is now a stable feature. The feature gates will be removed in 1.21 release. For more information, refer to notes below on the changelogs.
+
 ## Urgent Upgrade Notes 
 
 ### (No, really, you MUST read this before you upgrade)
 
 - A bug was fixed in kubelet where exec probe timeouts were not respected. Ensure that pods relying on this behavior are updated to correctly handle probe timeouts.
   
   This change in behavior may be unexpected for some clusters and can be disabled by turning off the ExecProbeTimeout feature gate. This gate will be locked and removed in future releases so that exec probe timeouts are always respected. ([#94115](https://github.com/kubernetes/kubernetes/pull/94115), [@andrewsykim](https://github.com/andrewsykim)) [SIG Node and Testing]
```
Comment (SergeyKanzhelev) [inline releases/release-1.20/release-notes-draft.md:83]: In response to [this thread](https://kubernetes.slack.com/archives/C2C40FMNF/p1607029719429000):

```suggestion
- A bug was fixed in kubelet where exec probe timeouts were not respected. This may result in unexpected behavior since the default timeout (if not specified) is `1s` which may be too small for some exec probes. Ensure that pods relying on this behavior are updated to correctly handle probe timeouts. See [configure probe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes) section of the documentation for more details.
  
  This change in behavior may be unexpected for some clusters and can be disabled by turning off the `ExecProbeTimeout` feature gate. This gate will be locked and removed in future releases so that exec probe timeouts are always respected. ([#94115](https://github.com/kubernetes/kubernetes/pull/94115), [@andrewsykim](https://github.com/andrewsykim)) [SIG Node and Testing]
```

Comment (SergeyKanzhelev) [reply releases/release-1.20/release-notes-draft.md:83]: Doc link will be updated with the 1.20 release of documentation 

https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes


will be

https://github.com/kubernetes/website/blob/dev-1.20/content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md#configure-probes

File: releases/release-1.20/release-notes-draft.md (hunk 4 above)
Comment (SergeyKanzhelev) [reply releases/release-1.20/release-notes-draft.md:182]: no deprecation warning if they are using beta. Alpha has deprecation warning for a few releases now. Good point though, maybe when we will place a deprecation warning it will be time to ask to take an action...

Comment (wilsonehusin) [reply releases/release-1.20/release-notes-draft.md:182]: okay yeah let's add this to deprecation warning and action required. I think going to v1 is quite a big deal that current users should start planning their upgrade strategy.

File: releases/release-1.20/release-notes-draft.md [hunk 6]
```diff
@@ -396,24 +493,39 @@
 
 ### Other (Cleanup or Flake)
```
Comment (sftim) [inline releases/release-1.20/release-notes-draft.md:495]: We're removing the documentation for the cloud-controller-manager skeleton binary (which doesn't actually integrate with any provider).

SIG API Machinery, is that correct? Slack discussion starts at https://kubernetes.slack.com/archives/C0EG7JC6T/p1606929363248600

File: releases/release-1.20/release-notes-draft.md [hunk 7]
```diff
@@ -1,178 +1,265 @@
...
+
+We expect this implementation to progress from alpha to beta and GA in coming releases, so we’re eager to have you comment about your dual-stack experiences in [#k8s-dual-stack](https://kubernetes.slack.com/messages/k8s-dual-stack) or in [enhancements #563](https://features.k8s.io/563). 
+
+### go1.15.2
+go1.15.2 has been integrated to Kubernets project as of this release, [including other infrastructure related updates on this effort](https://github.com/kubernetes/kubernetes/pull/94449).
+
+### CSI Volume Snapshot graduates to General Availability
+CSI Volume Snapshot moves to GA in the 1.20 release. This feature provides a standard way to trigger volume snapshot operations in Kubernetes and allows Kubernetes users to incorporate snapshot operations in a portable manner on any Kubernetes environment and supporting storage providers.
```
Comment (xing-yang) [inline releases/release-1.20/release-notes-draft.md:33]: on any Kubernetes environment and supporting storage providers -> on any Kubernetes environment regardless of the underlying storage

File: releases/release-1.20/release-notes-draft.md [hunk 8]
```diff
@@ -1,178 +1,265 @@
...
+
+### go1.15.2
+go1.15.2 has been integrated to Kubernets project as of this release, [including other infrastructure related updates on this effort](https://github.com/kubernetes/kubernetes/pull/94449).
+
+### CSI Volume Snapshot graduates to General Availability
+CSI Volume Snapshot moves to GA in the 1.20 release. This feature provides a standard way to trigger volume snapshot operations in Kubernetes and allows Kubernetes users to incorporate snapshot operations in a portable manner on any Kubernetes environment and supporting storage providers. 
+Additionally, these Kubernetes snapshot primitives act as basic building blocks that unlock the ability to develop advanced, enterprise grade, storage administration features for Kubernetes: including application or cluster level backup solutions.
+Note that snapshot support will require Kubernetes distributors to bundle the Snapshot controller, Snapshot CRDs, and validation webhook. In addition, a CSI driver supporting the snapshot functionality must also be deployed on the cluster.
+
+### Non-recursive Volume Ownership (FSGroup) graduates to Beta
```
Comment (xing-yang) [inline releases/release-1.20/release-notes-draft.md:37]: This entire section is wrong.  It should be the following instead.

```
Non-recursive volume ownership (fsgroup) feature is promoted to beta in 1.20.

By default, the fsgroup setting, if specified, recursively updates permissions for every file in a volume on every mount. This can make mount, and pod startup, very slow if the volume has many files. 

This setting enables a pod to specify a PodFSGroupChangePolicy that indicates that volume ownership and permissions will be changed only when permission and ownership of the root directory does not match with expected permissions on the volume.
```

Comment (wilsonehusin) [reply releases/release-1.20/release-notes-draft.md:37]: oops my bad, must have been a dirty copy-paste

File: releases/release-1.20/release-notes-draft.md (hunk 6 above)
Comment (wilsonehusin) [reply releases/release-1.20/release-notes-draft.md:494]: I'm going to follow the conversation -- thanks for flagging Tim!

File: releases/release-1.20/release-notes-draft.md [hunk 9]
```diff
@@ -1,178 +1,264 @@
...
+
+### Introducing Graceful Node Shutdown (Alpha)
+The `GracefulNodeShutdown` feature is now in Alpha. This allows kubelet to be aware of node system shutdowns, enabling graceful termination of pods during a system shutdown. This feature can be [enabled through feature gate](https://docs.k8s.io/concepts/architecture/nodes/#graceful-node-shutdown).
+
+### Runtime log sanitation
+Logs can now be configured to use runtime protection from leaking sensitive data. [Details for this experimental feature is available in documentation](https://docs.k8s.io/concepts/cluster-administration/system-logs/#log-sanitization).
+
+### Pod resource metrics
+On-demand metrics calculation is now available through `/metrics/resources`. [When enabled](
+https://docs.k8s.io/concepts/cluster-administration/system-metrics#kube-scheduler-metrics), the endpoint will report the requested resources and the desired limits of all running pods.
```
Comment (ehashman) [inline releases/release-1.20/release-notes-draft.md:55]: SIG Instrumentation notes LGTM.

Since doc changes aren't live yet, links for reference:

- https://deploy-preview-24845--kubernetes-io-vnext-staging.netlify.app/docs/concepts/cluster-administration/system-logs/#log-sanitization
- https://deploy-preview-24910--kubernetes-io-vnext-staging.netlify.app/docs/concepts/cluster-administration/system-metrics/#kube-scheduler-metrics

**Review by ehashman — APPROVED** [review]

File: releases/release-1.20/release-notes-draft.md [hunk 10]
```diff
@@ -1,178 +1,264 @@
...
+
+### Dockershim deprecation
+Docker as an underlying runtime is being deprecated. Docker-produced images will continue to work in your cluster with all runtimes, as they always have.
+The Kubernetes community [has written a blog post about this in detail](https://blog.k8s.io/2020/12/02/dont-panic-kubernetes-and-docker/) with [a dedicated FAQ page for it](https://blog.k8s.io/2020/12/02/dockershim-faq/).
+
+### External credential provider for client-go
+The client-go credential plugins can now be passed in the current cluster information via the `KUBERNETES_EXEC_INFO` environment variable. Learn more about this on [client-go credential plugins documentation](https://docs.k8s.io/reference/access-authn-authz/authentication/#client-go-credential-plugins/).
+
+### CronJob controller v2 is available through feature gate
+An alternative implementation of `CronJob` controller is now available as an alpha feature in this release, which has expermiental performance improvement by using informers instead of polling. While this will be the default behavior in the future, you can [try them in this release through a feature gate](https://docs.k8s.io/concepts/workloads/controllers/cron-jobs/).
```
Comment (SergeyKanzhelev) [inline releases/release-1.20/release-notes-draft.md:16]:
```suggestion
An alternative implementation of `CronJob` controller is now available as an alpha feature in this release, which has experimental performance improvement by using informers instead of polling. While this will be the default behavior in the future, you can [try them in this release through a feature gate](https://docs.k8s.io/concepts/workloads/controllers/cron-jobs/).
```

File: releases/release-1.20/release-notes-draft.md [hunk 11]
```diff
@@ -1,178 +1,264 @@
...
+Troubleshoot distroless containers by adding a new container with debugging tools, either in a new copy of the pod or using an ephemeral container. (Ephemeral containers are an alpha feature that are not enabled by default.)
+Troubleshoot on a node by creating a container running in the host namespaces and with access to the host’s filesystem.
+Note that as a new builtin command, `kubectl debug` takes priority over any `kubectl` plugin named “debug”. You will need to rename the affected plugin.
+Invocations using `kubectl alpha debug` are now deprecated and will be removed in a subsequent release. Update your scripts to use `kubectl debug` instead of `kubectl alpha debug`!
+For more information about kubectl debug, see Debugging Running Pods on the Kubernetes website, kubectl help debug, or reach out to SIG CLI by visiting #sig-cli or commenting on [enhancement #1441](https://features.k8s.io/1441).
+
+### Removing deprecated flags in kubeadm
+`kubeadm` removes a significant number of deprecated flags in this release. The complete list of flags are available below on Urgent Upgrade Notes section.
```
Comment (neolit123) [inline releases/release-1.20/release-notes-draft.md:71]: i'd reword this section to something in the lines of the following:

```
### Deprecations and deprecated feature removals in kubeadm
`kubeadm` applies a number of deprecations and removals of deprecated features in this release. More details are available in the Urgent Upgrade Notes and Kind / Deprecation sections.
```

rationale:
- the most important deprecation is the "master" label/taint.
- the deprecations and removals are spread across a couple of different sections.

**Comment by liggitt** [comment]
The release note for https://github.com/kubernetes/kubernetes/pull/92743 should be hoisted to the "before upgrade" section. I just updated the release note to include a link to a tool that can be run prior to upgrade to identify existing objects with invalid references.

> The [kubectl-check-ownerreferences](https://github.com/kubernetes-sigs/kubectl-check-ownerreferences) tool can be run prior to upgrading to locate existing objects with invalid ownerReferences.

File: releases/release-1.20/release-notes-draft.md (hunk 11 above)
Comment (wilsonehusin) [reply releases/release-1.20/release-notes-draft.md:71]: sounds good to me, thanks for reviewing @neolit123!

File: releases/release-1.20/release-notes-draft.md [hunk 12]
```diff
@@ -1,178 +1,264 @@
-# Release notes for v1.20.0-beta.2
+# Release notes for v1.20.0-rc.0
 
 [Documentation](https://docs.k8s.io/docs/home)
 # Changelog since v1.19.0
 
+## What's New (Major Themes)
+
+### Dockershim deprecation
```
Comment (celestehorgan) [inline releases/release-1.20/release-notes-draft.md:8]: fyi @wilsonehusin – when these eventually make their way over to `k/website`, each of these headers will need a full line underneath them, or the header rendering will break:

```suggestion
### Dockershim deprecation

```

Comment (wilsonehusin) [reply releases/release-1.20/release-notes-draft.md:8]: 😭 okay I'll get on it, thanks Celeste!

File: releases/release-1.20/release-notes-draft.md [hunk 13]
```diff
@@ -1,178 +1,293 @@
...
+
+### IPv4/IPv6 run
+
+IPv4/IPv6 dual-stack has been reimplemented for 1.20 to support dual-stack Services, based on user and community feedback. If your cluster has dual-stack enabled, you can create Services which can use IPv4, IPv6, or both, and you can change this setting for existing Services. Details are available in updated [IPv4/IPv6 dual-stack docs](https://docs.k8s.io/concepts/services-networking/dual-stack/), which cover the nuanced array of options.
+
+We expect this implementation to progress from alpha to beta and GA in coming releases, so we’re eager to have you comment about your dual-stack experiences in [#k8s-dual-stack](https://kubernetes.slack.com/messages/k8s-dual-stack) or in [enhancements #563](https://features.k8s.io/563). 
+
+### go1.15.2
+
+go1.15.2 has been integrated to Kubernets project as of this release, [including other infrastructure related updates on this effort](https://github.com/kubernetes/kubernetes/pull/94449).
```
Comment (xmudrii) [inline releases/release-1.20/release-notes-draft.md:38]: @wilsonehusin The 1.20.0 release will be built with Go 1.15.5 (ref: https://github.com/kubernetes/kubernetes/pull/95776). I'm not sure is the migration fully-done yet, but should we mention go1.15.2 or go1.15.5 here?

cc @justaugustus -- maybe you have some input on this. :slightly_smiling_face:

Comment (wilsonehusin) [reply releases/release-1.20/release-notes-draft.md]: I'm updating it as 1.15.5 per https://github.com/kubernetes/release/issues/1651 (checkboxes for 1.15.5 are all ✅ )

Comment (wilsonehusin) [reply releases/release-1.20/release-notes-draft.md]: thanks for bringing it up @xmudrii!

**Comment by wilsonehusin** [comment]
alright thank you so much everyone for the cooperation!

/hold cancel

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Comment by puerco** [comment]
Great job @wilsonehusin . Thank you very much !!
